io/src/tim_grabber.cpp

```diff
@@ -0,0 +1,210 @@
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::initialize ()
+{
+  buildLookupTables ();
+
+  point_cloud_xyz_signal_ = createSignal<sig_cb_sick_tim_scan_point_cloud_xyz> ();
+
+  point_cloud_xyz_ptr_.reset (new pcl::PointCloud<pcl::PointXYZ>);
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+float
+pcl::TimGrabber::getFramesPerSecond () const
+{
+  boost::mutex::scoped_lock lock( frequency_mutex_ );
+  return (frequency_.getFrequency ());
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::buildLookupTables () {
+  cos_dynamic_lookup_table_.resize (amount_of_data_);
+  sin_dynamic_lookup_table_.resize (amount_of_data_);
+
+  const float angle_step = angle_range_ / static_cast<float>(amount_of_data_);
+  float angle = angle_start_;
+
+  for (std::size_t i = 0; i < amount_of_data_; i++)
+  {
+    cos_dynamic_lookup_table_[i] = std::cos (angle);
+    sin_dynamic_lookup_table_[i] = std::sin (angle);
+    angle += angle_step;
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::updateLookupTables () {
+  //// note: cos, sin are the same size
+  if (cos_dynamic_lookup_table_.size () != amount_of_data_)
+    buildLookupTables ();
+}
+
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+bool
+pcl::TimGrabber::isValidPacket () const {
+  return received_packet_.data ()[length_-1] == '\03';
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::receiveTimPacket ()
+{
+  if (!tim_socket_.is_open ())
+    return;
+
+  tim_socket_.send (boost::asio::buffer ("\x02sRN LMDscandata 1\x03\0"));
+
+  std::this_thread::sleep_for (std::chrono::milliseconds (wait_time_milliseconds_));
+
+  length_ = tim_socket_.receive (boost::asio::buffer (received_packet_));
+
+  if (!isValidPacket ()) {
+    //// Countup wait_time
+    //// Because fit wait_time to network latency
+    wait_time_milliseconds_++;
+
+    //// Clear invalid packet in socket
+    while (!isValidPacket ())
+      length_ = tim_socket_.receive (boost::asio::buffer (received_packet_));
+
+    //// If received packet is invalid, recurse
+    receiveTimPacket ();
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::parseTimPacket ()
+{
+  std::stringstream ss (received_packet_.data ());
+  std::string str;
+
+  for (int i=0; i<26; ++i)
+    ss >> str;
+
+  amount_of_data_ = std::stoi (str, nullptr, 16);
+
+  point_cloud_xyz_ptr_->resize (amount_of_data_);
+
+  distances_.resize (amount_of_data_);
+  for (auto& distance : distances_) {
+    ss >> str;
```

Comment:
Wil the following work?
```
ss.setf(std::hex);
for (auto& distance:distances_) {
  ss >> distance;
  distance /= 1000.0;  // convert from [mm] to [m]
}
```

Reply:
Did you try `ss >> std::hex >> distance`? (Also, it has to be `setf` not `self` (T not L))

Reply:
Sorry, I got a typo when write comment.

Yes, I tried it (and retry it now). But it does not work.

It seems that std :: hex cannot be streamed to float type
So once I put a variable of type ʻint`, it worked.

```
int d;
for (auto& distance : distances_) {
  ss >> std::hex >> d;
  distance = d / 1000.0;
}
```

Thank you!